clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -292,15 +292,26 @@ public void close() {
      */
     public void send(Send send) {
         String connectionId = send.destination();
-        if (closingChannels.containsKey(connectionId))
+        KafkaChannel channel = channelOrFail(connectionId);
+        if (closingChannels.containsKey(connectionId)) {
+            channel.state(ChannelState.FAILED_SEND);
+            // Ensure notification via `disconnected`
             this.failedSends.add(connectionId);
+        }
         else {
-            KafkaChannel channel = channelOrFail(connectionId, false);
             try {
                 channel.setSend(send);
-            } catch (CancelledKeyException e) {
+            } catch (Exception e) {
+                // Update the state for consistency, the channel will be discarded after `close`
+                channel.state(ChannelState.FAILED_SEND);
+                // Ensure notification via `disconnected`
                 this.failedSends.add(connectionId);
                 close(channel, false);
+                if (!(e instanceof CancelledKeyException)) {
+                    log.error("Unexpected exception during send, closing connection {} and rethrowing exception {}",
+                            connectionId, e);
+                    throw e;
```

Comment:
Do we think it makes sense to rethrow the exception here? Since this should only happen due to bugs, I think it's probably good not to swallow the exception, but would be interested in other opinions.